### ISSUE TITLE
Parallel shift

### ISSUE THREAD

**Issue by mlerjen** [OpenOrienteering/mapper#650]
![paralell shift](https://cloud.githubusercontent.com/assets/15611653/14310061/328b16f4-fbe0-11e5-8e98-cd44b5239825.jpg)
To draw exactly parallel lines (as a fence following along a winding road) it would be nice to have a parallel shift function ~ moving the points of a line symbol perpendicular to their tangents (?).

The parallel fence then could be drawn by "filling" the road with the fence first and then using that parallel shift...

**Comment by wanacode** [comment]
This sounds similar to this issue [https://github.com/OpenOrienteering/mapper/issues/362](https://github.com/OpenOrienteering/mapper/issues/362)

It would definitely be useful. I would be interested to hear any clever ways of drawing parallel lines in the mean time.

**Comment by dg0yt** [comment]
Another variant of #143.

**Comment by yevhenmazur** [comment]
Please note - this kind of transformation called Concentric Shift or Radial Shift. The Parallel shift already implemented as Duplicate + Move Object. 
![concentric_shift](https://cloud.githubusercontent.com/assets/10380431/17079245/92ab1674-5112-11e6-9671-c848b1c7f872.png)

**Comment by mlerjen** [comment]
The new "follow outline" feature opens up for a temporary work-around to this concentric-shift issue... https://youtu.be/0nXfyCYcHjM

**Comment by Symbian9** [comment]
There is new open-source lib `CavalierContours` (MIT-licensed) that could be useful for *Mapper*, but as for now it limited only to *"processing 2D polylines containing both straight line and constant radius arc segments"* (no Bezier curves support yet)

- Lib (core) - https://github.com/jbuckmccready/CavalierContours
- GUI (interactive) - https://github.com/jbuckmccready/CavalierContoursDev

![pic.1](https://raw.githubusercontent.com/jbuckmccready/CavalierContoursDoc/master/images/pretty_examples/example1.png)

![pic.2](https://raw.githubusercontent.com/jbuckmccready/CavalierContoursDoc/master/images/pretty_examples/example6.png)

**Comment by dg0yt** [comment]
> There is new open-source lib `CavalierContours` (MIT-licensed) that could be useful for _Mapper_, but as for now it limited only to _"processing 2D polylines containing both straight line and constant radius arc segments"_ (no Bezier curves support yet)

There may be more libraries, but we already do have `QBezier` for such calculations. Someone has to write the code which uses it.

**Comment by lpechacek** [comment]
> Someone has to write the code which uses it.

It's on my list and I've already made one (failed) attempt to implement it as a standalone tool. However, I'm more and more sure that it essentially should be a function of the Edit Line tool, which moves **one segment** in a "concentric" fashion. It's natural to extend the tool to move *all* path segments in a similar way. I'm wondering how the function should be switched. Any suggestions for modifier keys and/or switch key combinations are welcome.

BTW, I need this function too. :)

**Comment by dg0yt** [comment]
Maybe better start with a standalone clone of the edit-line tool: I would prefer to select multiple bezier segments by dragging along the line, before shifting by dragging away from the line.

**Comment by dg0yt** [comment]
> Thus I would prefer that the default is "all segments of the selected objects", with the option to deselect sections by ctrl-clicking on them before draging away.

One more reason to start with a separate tool ;-)

**Comment by lpechacek** [comment]
> One more reason to start with a separate tool ;-)

Yeah, this was a good choice. Thanks for the hint, Kai. There is a mostly working implementation is in my repo (see above). The bad thing is that either `LineSymbol::shiftCoordinates()` needs a substantial improvement or we need alternate implementation for the radial path shift. The path shape quickly degrades for large shift distances.

**Comment by jmacura** [comment]
Just a little important note about naming of this great future tool: _parallel curve_ or, perhaps better, _offset curve_ seems to be adequate. See https://en.wikipedia.org/wiki/Parallel_curve

**Comment by mlerjen** [comment]
@jmacura _offset line_ maybe? 
![image](https://user-images.githubusercontent.com/15611653/112665799-baacff80-8e5b-11eb-9306-0cbdb700cdd3.png)

**Comment by dl3sdo** [comment]
@lpechacek: thank you for providing an implementation for that topic, Libor. I'm not sure how to use it correctly as it crashes Mapper (Debug mode) everytime when I apply it twice on the same line object (I select the tool, move a line, release the mouse button, move it again -> crash (crash on moving the mouse)).

**Comment by lpechacek** [comment]
> ... it crashes Mapper (Debug mode) everytime when I apply it twice on the same line object.

Thanks for the early testing, Matthias. I appreciate your interest in the feature.

I'm aware that the tool crashes under certain condition but your specific case worked in my testing today. Further code cleanup and testing is on my schedule. In any case, you are welcome to analyze the crash. I guess that it will be null pointer dereference on `hover_object` in `dragMove()` or something similar.

**Comment by jmacura** [comment]
> @jmacura _offset line_ maybe? 
> ![image](https://user-images.githubusercontent.com/15611653/112665799-baacff80-8e5b-11eb-9306-0cbdb700cdd3.png)
> 

Mathematicaly, a _line_ is a straight segment of 2D space. But fine, consistent naming should be used across the GUI. I can't check right ow, but if there is a _draw line_, than _offset line_ makes complete sense.

**Comment by lpechacek** [comment]
> Just a little important note about naming of this great future tool: ...

While deciding about the name, I've started with Yevhen's comment from July 2016 above. As far as I remember, I also googled a bit about the terminology, and it made sense to me when I took the ticket.

In any case, I prefer an established name ("Move parallel" is the OCAD tool name since immemorial) or a name from the GIS world or cartography, which are areas that relate to o-map making as far as I can tell. Just agree on the name here, and I'll implement the code change. Thanks!

**Comment by dl3sdo** [comment]
I'm in favour of "Move parallel" as being most intuitive (and thus serving its purpose), in addition refer to https://en.wikipedia.org/wiki/Parallel_curve

**Comment by ollesmaps** [comment]
"Move parallel" is fine. Offset might be logicaly suitable however in the orienteering context it relates more to printing.

**Comment by ollesmaps** [comment]
On a sidenote, would be possible to use this tool (or a new one?) like this:
1. acquired data of buildings for a sprint map 
2. a lot of these buildings touch each other
3. goal: seamlessly merge these buildings, so there are no seams
4. my intuitive solution (in Ocad) was to select all buildings
5. enlarge lets say by 10 percent
6. merge all these buildings
7. decrease size by 9,0909... percent

However the resulted vector is not what anticipated. It does not work. I assume a "higher" mathematic is involved.
Such a tool would be very much apprwciated though.

**Comment by lpechacek** [comment]
>     4. my intuitive solution (in Ocad) was to select all buildings
> 
>     5. enlarge lets say by 10 percent

In Mapper: Select all buildings, activate Scale tool (Z), press and hold Ctrl to scale selected objects around their centre (see #1211 for a screenshot), enlarge buildings to 110%.

>     6. merge all these buildings

Activate Unify areas tool (press U).

>     7. decrease size by 9,0909... percent

Press and hold Ctrl and drag cursor to reduce buildings to 90.9%. HTH

> However the resulted vector is not what anticipated. It does not work.

Without knowledge of the expectation, I cannot comment on this part. In any case, I don't think that this use case (merge building areas) requires cooperation from the new tool.

**Comment by ollesmaps** [comment]
Thank you Libor for the pointer.
I tried the approach - very powerful indeed.

However I am not able to get the desired result.  (But I can be doing something wrong. Will further explore it.)
See the attached picture:
Pic 1: situation
Pic 2: desired result
Pic 3: what I get (green color) when I use Scale tool + Ctrl

![buildings_merge](https://user-images.githubusercontent.com/20535899/113001573-374c2080-9171-11eb-8e35-2af2919f9502.jpg)

**Comment by lpechacek** [comment]
I think that I see what you mean. The shape change resulting from the scaling around the new object centre is undesirable. In my opinion, this is a new use case and should be reported as a new enhancement request.

**Comment by mlerjen** [comment]
I noticed that using the pan tool, the offset and distance is shown in the status bar. It would be very nice to find similar information  about the parallel move there (the parallel distance).

**Comment by lpechacek** [comment]
The distance indication is in:
`setStatusBarText(tr("<b>Move distance:</b> %1 mm").arg(QLocale().toString(qFabs(move_distance), 'f', 1)));`
I added it the day before yesterday.
